### DISCUSSION TITLE
Wait with initialization or loading of languages until after authorization

### DISCUSSION THREAD

**Discussion by silvio-e** [i18next/i18next#1935] [Q&A]
Hello!

In our React application, I need to fetch translations from endpoints for which the user needs to be authenticated first so that we can request tokens for these endpoints and then get the translations.

I tried without luck to move the initialization from the `main.tsx` into the `App.tsx` so the user is authenticated and into a `useEffect` to wait for other things that happen async. But it seems that i18n gets initialized twice.

Here the hook and initialization:

```js
useEffect(() => {
  if (!i18n.isInitialized && env) {
    i18n
      .use(Backend)
      .use(initReactI18next)
      .init({
        backend: {
          loadPath: `https://my-json-server.typicode.com/silvio-e/translations-test/translations?test=${env.VITE_ENVIRONMENT}`
        },
        returnNull: false,
        fallbackLng: "en",
        debug: true,
        interpolation: {
          escapeValue: false // not needed for react as it escapes by default
        }
      });
  }
}, [env]);
```

But this would result in the following warnings:

```
⚠️ react-i18next:: You will need to pass in an i18next instance by using initReactI18next
i18next: initialized ...
GET https://my-json-server.typicode.com/silvio-e/translations-test/translations?test=dev
i18next::backendConnector: loaded namespace translation for language en 
Object { "Find-customer": "Hello world!" }
[i18next.js:25:48](http://localhost:3000/node_modules/i18next/dist/esm/i18next.js)
i18next: languageChanged en
⚠️ i18next: init: i18next is already initialized. You should call init just once!
i18next: initialized
```

It loads the translation but can't rerender and apply it.

I can't find anything in the documentation neither in i18next, react-i18next or i18next-http-backend. Does anyone have any advice on how I could achieve my goal here?

**Comment by jamuhl** [comment]
`i18n.isInitialized` is set when initialization is done...so if your use effect runs before finishing again...you run into double init

better set you own flag that you called the init function


what I would do:

- init i18next as normally done (without useEffect)
- use a HOC that asserts user is authenticated -> throw a promise triggering suspense (resolve on authenticated)

that way pages and call to useTranslation are delayed until authentication is ready
